File: packages/game-rules/src/state-hash.ts
/**
 * WP-030: Deterministic TS NLHE engine state hash (differential oracle).
 *
 * This is NOT Protocol V3 event encoding (`MOZETTO_POKER_EVENT_V1`) and NOT the
 * Protocol V3 `engineHash` build id. See docs/WP-030_TS_ENGINE_FREEZE.md.
 */
import { concat, keccak256, toBytes, toHex, type Hex } from "viem";
import { cardKey } from "./cards.js";
import type { HoldemState, LegalAction, SeatState } from "./holdem.js";
import { chipsToNumber, type Chips } from "./money.js";

/** Domain string for the TS freeze-oracle state digest. */
export const TS_ENGINE_STATE_DOMAIN = "MOZETTO_TS_ENGINE_STATE_V1" as const;

/**
 * Build identity for the TypeScript NLHE engine (NLHE_ENGINE_RC1).
 */
export const TS_ENGINE_BUILD_ID = "mozetto-nlhe-ts-rc1" as const;

/** Protocol V3 Season 1 draft placeholder (specs/MOZETTO_POKER_EVENT_V1.md §9 — frozen). */
export const PROTOCOL_V3_ENGINE_HASH_PLACEHOLDER_ID = "mozetto-nlhe-engine-v3-draft" as const;

/**
 * GameTemplate.engineHash promotion target (NLHE_ENGINE_RC1).
 */
export const GAME_TEMPLATE_ENGINE_BUILD_ID = "mozetto-nlhe-engine-rc1" as const;

function chipNum(v: Chips | number | null | undefined): number | null {
  if (v == null) return null;
  return chipsToNumber(typeof v === "bigint" ? v : BigInt(Math.trunc(v)));
}

export function tsEngineBuildHash(): Hex {
  return keccak256(toBytes(TS_ENGINE_BUILD_ID));
}

export function protocolV3EngineHashPlaceholder(): Hex {
  return keccak256(toBytes(PROTOCOL_V3_ENGINE_HASH_PLACEHOLDER_ID));
}

export function gameTemplateEngineHash(): Hex {
  return keccak256(toBytes(GAME_TEMPLATE_ENGINE_BUILD_ID));
}

export type ConsensusSeat = {
  seatIndex: number;
  stack: number;
  bet: number;
  totalBet: number;
  hole: string[] | null;
  folded: boolean;
  allIn: boolean;
  sitOut: boolean;
  /** Occupancy marker only (empty string when vacant). */
  occupied: boolean;
};

/**
 * Consensus snapshot: betting/card legality fields only.
 * Excludes: serverSeed, playerId, agentId, display labels on winners.
 */
export type ConsensusSnapshot = {
  domain: typeof TS_ENGINE_STATE_DOMAIN;
  buildId: typeof TS_ENGINE_BUILD_ID;
  config: {
    tableId: string;
    smallBlind: number;
    bigBlind: number;
    rakePct: number;
    rakeCap: number | null;
  };
  handId: string | null;
  handNumber: number;
  street: HoldemState["street"];
  button: number;
  /** Remaining undealt deck (card keys, top = index 0). */
  deck: string[];
  board: string[];
  pot: number;
  seats: ConsensusSeat[];
  actingIndex: number | null;
  currentBet: number;
  minRaise: number;
  lastAggressor: number | null;
  firstToAct: number | null;
  seedCommit: string | null;
  winners: { seatIndex: number; amount: number }[];
  rake: number;
  actedThisStreet: number[];
  lastRaiseComplete: boolean;
};

function seatSnapshot(s: SeatState): ConsensusSeat {
  return {
    seatIndex: s.seatIndex,
    stack: chipsToNumber(s.stack),
    bet: chipsToNumber(s.bet),
    totalBet: chipsToNumber(s.totalBet),
    hole: s.hole ? s.hole.map(cardKey) : null,
    folded: s.folded,
    allIn: s.allIn,
    sitOut: s.sitOut,
    occupied: Boolean(s.playerId),
  };
}

/** Stable JSON with sorted object keys (arrays keep order). */
export function stableStringify(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((v) => stableStringify(v)).join(",")}]`;
  }
  const obj = value as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${stableStringify(obj[k])}`).join(",")}}`;
}

export function toConsensusSnapshot(state: HoldemState): ConsensusSnapshot {
  return {
    domain: TS_ENGINE_STATE_DOMAIN,
    buildId: TS_ENGINE_BUILD_ID,
    config: {
      tableId: state.config.tableId,
      smallBlind: chipsToNumber(state.config.smallBlind),
      bigBlind: chipsToNumber(state.config.bigBlind),
      rakePct: state.config.rakePct,
      rakeCap: chipNum(state.config.rakeCap),
    },
    handId: state.handId,
    handNumber: state.handNumber,
    street: state.street,
    button: state.button,
    deck: state.deck.map(cardKey),
    board: state.board.map(cardKey),
    pot: chipsToNumber(state.pot),
    seats: [...state.seats].sort((a, b) => a.seatIndex - b.seatIndex).map(seatSnapshot),
    actingIndex: state.actingIndex,
    currentBet: chipsToNumber(state.currentBet),
    minRaise: chipsToNumber(state.minRaise),
    lastAggressor: state.lastAggressor,
    firstToAct: state.firstToAct,
    seedCommit: state.seedCommit,
    winners: state.winners.map((w) => ({ seatIndex: w.seatIndex, amount: chipsToNumber(w.amount) })),
    rake: chipsToNumber(state.rake),
    actedThisStreet: [...state.actedThisStreet].sort((a, b) => a - b),
    lastRaiseComplete: state.lastRaiseComplete,
  };
}

/**
 * keccak256( domainTag || keccak256(utf8(stable JSON snapshot)) ).
 * Domain tag = keccak256(bytes(TS_ENGINE_STATE_DOMAIN)).
 *
 * Documented interim: Protocol V3 forbids raw JSON for consensus object hashes;
 * this freeze oracle uses JSON only as the state preimage under an explicit
 * TS-engine domain until a formal ENGINE_STATE ABI layout is specified.
 */
export function hashEngineState(state: HoldemState): Hex {
  const snapshot = toConsensusSnapshot(state);
  const body = keccak256(toBytes(stableStringify(snapshot)));
  const domainTag = keccak256(toBytes(TS_ENGINE_STATE_DOMAIN));
  return keccak256(concat([domainTag, body]));
}

export function hashLegalActions(actions: LegalAction[]): Hex {
  const normalized = actions
    .map((a) => ({
      action: a.action,
      minAmount: a.minAmount != null ? chipsToNumber(a.minAmount) : null,
      maxAmount: a.maxAmount != null ? chipsToNumber(a.maxAmount) : null,
    }))
    .sort((a, b) => a.action.localeCompare(b.action));
  const domainTag = keccak256(toBytes(`${TS_ENGINE_STATE_DOMAIN}:legal`));
  const body = keccak256(toBytes(stableStringify(normalized)));
  return keccak256(concat([domainTag, body]));
}

/** Hex digest of utf8 bytes — helper for fixture metadata. */
export function digestUtf8(s: string): Hex {
  return keccak256(toBytes(s));
}

export function assertHexEqual(actual: Hex, expected: string, label: string): void {
  const norm = expected.startsWith("0x") ? expected.toLowerCase() : `0x${expected}`.toLowerCase();
  if (actual.toLowerCase() !== norm) {
    throw new Error(`${label}: expected ${norm}, got ${actual}`);
  }
}

export function snapshotStacks(state: HoldemState): number[] {
  return [...state.seats].sort((a, b) => a.seatIndex - b.seatIndex).map((s) => chipsToNumber(s.stack));
}

export function formatHash(h: Hex): string {
  return toHex(h);
}

File: crates/poker-core/src/state_hash.rs
//! TS freeze-oracle state hash (`MOZETTO_TS_ENGINE_STATE_V1`).
//! Matches `packages/game-rules/src/state-hash.ts` for differential parity.

use crate::legal::LegalAction;
use crate::types::{HoldemState, TableConfig};
use alloy_primitives::{keccak256, B256};
use poker_eval::card_key;
use std::fmt::Write as _;

pub const TS_ENGINE_STATE_DOMAIN: &str = "MOZETTO_TS_ENGINE_STATE_V1";
pub const TS_ENGINE_BUILD_ID: &str = "mozetto-nlhe-ts-freeze-wp030";

/// Format rake_bps as the TS JSON number for `rakePct` (e.g. 500 → `0.05`, 0 → `0`).
pub fn format_rake_pct(bps: u32) -> String {
    if bps == 0 {
        return "0".into();
    }
    // bps / 10000 as decimal without scientific notation; trim trailing zeros.
    let whole = bps / 10_000;
    let frac = bps % 10_000;
    let mut s = format!("{whole}.{frac:04}");
    while s.ends_with('0') {
        s.pop();
    }
    if s.ends_with('.') {
        s.pop();
    }
    s
}

fn stable_stringify_null() -> &'static str {
    "null"
}

fn escape_json_string(s: &str) -> String {
    serde_json::to_string(s).expect("string serialize")
}

/// Stable JSON with sorted object keys (arrays keep order) — mirrors TS `stableStringify`.
pub fn stable_stringify_value(v: &serde_json::Value) -> String {
    match v {
        serde_json::Value::Null => stable_stringify_null().into(),
        serde_json::Value::Bool(b) => if *b { "true" } else { "false" }.into(),
        serde_json::Value::Number(n) => n.to_string(),
        serde_json::Value::String(s) => escape_json_string(s),
        serde_json::Value::Array(arr) => {
            let parts: Vec<String> = arr.iter().map(stable_stringify_value).collect();
            format!("[{}]", parts.join(","))
        }
        serde_json::Value::Object(map) => {
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            let parts: Vec<String> = keys
                .into_iter()
                .map(|k| {
                    format!(
                        "{}:{}",
                        escape_json_string(k),
                        stable_stringify_value(&map[k])
                    )
                })
                .collect();
            format!("{{{}}}", parts.join(","))
        }
    }
}

fn config_json(cfg: &TableConfig) -> serde_json::Value {
    let mut obj = serde_json::Map::new();
    obj.insert("tableId".into(), cfg.table_id.clone().into());
    obj.insert("smallBlind".into(), cfg.small_blind.into());
    obj.insert("bigBlind".into(), cfg.big_blind.into());
    // Emit rakePct as JSON number matching TS (not an integer bps field).
    let pct = format_rake_pct(cfg.rake_bps);
    let num: serde_json::Number = pct.parse().expect("rake pct number");
    obj.insert("rakePct".into(), serde_json::Value::Number(num));
    obj.insert(
        "rakeCap".into(),
        match cfg.rake_cap {
            Some(c) => c.into(),
            None => serde_json::Value::Null,
        },
    );
    serde_json::Value::Object(obj)
}

pub fn to_consensus_snapshot(state: &HoldemState) -> serde_json::Value {
    let mut seats: Vec<_> = state.seats.iter().collect();
    seats.sort_by_key(|s| s.seat_index);

    let seat_vals: Vec<serde_json::Value> = seats
        .into_iter()
        .map(|s| {
            let hole = match &s.hole {
                Some(h) => serde_json::Value::Array(
                    h.iter()
                        .map(|c| serde_json::Value::String(card_key(*c)))
                        .collect(),
                ),
                None => serde_json::Value::Null,
            };
            let mut o = serde_json::Map::new();
            o.insert("seatIndex".into(), s.seat_index.into());
            o.insert("stack".into(), s.stack.into());
            o.insert("bet".into(), s.bet.into());
            o.insert("totalBet".into(), s.total_bet.into());
            o.insert("hole".into(), hole);
            o.insert("folded".into(), s.folded.into());
            o.insert("allIn".into(), s.all_in.into());
            o.insert("sitOut".into(), s.sit_out.into());
            o.insert("occupied".into(), (!s.player_id.is_empty()).into());
            serde_json::Value::Object(o)
        })
        .collect();

    let winners: Vec<serde_json::Value> = state
        .winners
        .iter()
        .map(|w| {
            let mut o = serde_json::Map::new();
            o.insert("seatIndex".into(), w.seat_index.into());
            o.insert("amount".into(), w.amount.into());
            serde_json::Value::Object(o)
        })
        .collect();

    let acted: Vec<serde_json::Value> = state
        .acted_this_street
        .iter()
        .map(|i| serde_json::Value::from(*i))
        .collect();

    let mut snap = serde_json::Map::new();
    snap.insert("domain".into(), TS_ENGINE_STATE_DOMAIN.into());
    snap.insert("buildId".into(), TS_ENGINE_BUILD_ID.into());
    snap.insert("config".into(), config_json(&state.config));
    snap.insert(
        "handId".into(),
        match &state.hand_id {
            Some(h) => h.clone().into(),
            None => serde_json::Value::Null,
        },
    );
    snap.insert("handNumber".into(), state.hand_number.into());
    snap.insert("street".into(), state.street.as_str().into());
    snap.insert("button".into(), state.button.into());
    snap.insert(
        "deck".into(),
        serde_json::Value::Array(
            state
                .deck
                .iter()
                .map(|c| serde_json::Value::String(card_key(*c)))
                .collect(),
        ),
    );
    snap.insert(
        "board".into(),
        serde_json::Value::Array(
            state
                .board
                .iter()
                .map(|c| serde_json::Value::String(card_key(*c)))
                .collect(),
        ),
    );
    snap.insert("pot".into(), state.pot.into());
    snap.insert("seats".into(), serde_json::Value::Array(seat_vals));
    snap.insert(
        "actingIndex".into(),
        match state.acting_index {
            Some(i) => i.into(),
            None => serde_json::Value::Null,
        },
    );
    snap.insert("currentBet".into(), state.current_bet.into());
    snap.insert("minRaise".into(), state.min_raise.into());
    snap.insert(
        "lastAggressor".into(),
        match state.last_aggressor {
            Some(i) => i.into(),
            None => serde_json::Value::Null,
        },
    );
    snap.insert(
        "firstToAct".into(),
        match state.first_to_act {
            Some(i) => i.into(),
            None => serde_json::Value::Null,
        },
    );
    snap.insert(
        "seedCommit".into(),
        match &state.seed_commit {
            Some(s) => s.clone().into(),
            None => serde_json::Value::Null,
        },
    );
    snap.insert("winners".into(), serde_json::Value::Array(winners));
    snap.insert("rake".into(), state.rake.into());
    snap.insert("actedThisStreet".into(), serde_json::Value::Array(acted));
    snap.insert("lastRaiseComplete".into(), state.last_raise_complete.into());
    serde_json::Value::Object(snap)
}

pub fn hash_engine_state(state: &HoldemState) -> B256 {
    let snapshot = to_consensus_snapshot(state);
    let body_str = stable_stringify_value(&snapshot);
    let body = keccak256(body_str.as_bytes());
    let domain_tag = keccak256(TS_ENGINE_STATE_DOMAIN.as_bytes());
    let mut concat = [0u8; 64];
    concat[..32].copy_from_slice(domain_tag.as_slice());
    concat[32..].copy_from_slice(body.as_slice());
    keccak256(concat)
}

pub fn hash_legal_actions(actions: &[LegalAction]) -> B256 {
    let mut normalized: Vec<serde_json::Value> = actions
        .iter()
        .map(|a| {
            let mut o = serde_json::Map::new();
            o.insert("action".into(), a.action.as_str().into());
            o.insert(
                "minAmount".into(),
                match a.min_amount {
                    Some(v) => v.into(),
                    None => serde_json::Value::Null,
                },
            );
            o.insert(
                "maxAmount".into(),
                match a.max_amount {
                    Some(v) => v.into(),
                    None => serde_json::Value::Null,
                },
            );
            serde_json::Value::Object(o)
        })
        .collect();
    normalized.sort_by(|a, b| {
        let aa = a["action"].as_str().unwrap_or("");
        let bb = b["action"].as_str().unwrap_or("");
        aa.cmp(bb)
    });
    let domain = format!("{TS_ENGINE_STATE_DOMAIN}:legal");
    let body_str = stable_stringify_value(&serde_json::Value::Array(normalized));
    let body = keccak256(body_str.as_bytes());
    let domain_tag = keccak256(domain.as_bytes());
    let mut concat = [0u8; 64];
    concat[..32].copy_from_slice(domain_tag.as_slice());
    concat[32..].copy_from_slice(body.as_slice());
    keccak256(concat)
}

pub fn format_hash(h: &B256) -> String {
    let mut s = String::with_capacity(66);
    s.push_str("0x");
    for b in h.as_slice() {
        let _ = write!(s, "{b:02x}");
    }
    s
}
